truss/secrets_manager.go

```diff
@@ -0,0 +1,355 @@
+	}
+	if _, err = vault.PortForward(); err != nil {
+		return false, err
+	}
+	defer vault.ClosePortForward()
+
+	// load existing disk value
+	// decrypt it or provide default
+	raw, err := m.GetDecryptedFromDisk(vault, environment)
+	if err != nil {
+		return false, err
+	}
+
+	// save to tmp file
+	tmpFile, err := ioutil.TempFile("", "trussvault-*")
+	if err != nil {
+		return false, err
+	}
+	defer os.Remove(tmpFile.Name())
+	tmpFile.Write(raw)
+	tmpFile.Close()
+
+	info, _ := os.Stat(tmpFile.Name())
+	modTimeAtOpen := info.ModTime()
+
+	// vim tmp file
+	cmd := exec.Command(viper.GetString("EDITOR"), tmpFile.Name())
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Run()
+
+	// check if saved
+	info, _ = os.Stat(tmpFile.Name())
+	modTimeAtClose := info.ModTime()
+	if !modTimeAtClose.After(modTimeAtOpen) {
+		return false, nil
+	}
+
+	// encrypt and save to disk
+	raw, err = ioutil.ReadFile(tmpFile.Name())
+	if err != nil {
+		return true, err
+	}
+	if err := m.EncryptAndSaveToDisk(vault, environment, raw); err != nil {
+		return true, err
+	}
+
+	return true, nil
+}
+
+// PushAll pushes all secrets for all environments
+func (m SecretsManager) PushAll() error {
+	for env := range m.Environments {
+		if err := m.Push(env); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Push pushes secrets to Vaut
+func (m SecretsManager) Push(environment string) error {
+	vault, err := m.Vault(environment)
+	if err != nil {
+		return err
+	}
+	if _, err := vault.PortForward(); err != nil {
+		return err
+	}
+	defer vault.ClosePortForward()
+
+	secrets, err := m.GetMapFromDisk(vault, environment)
+	if err != nil {
+		return err
+	}
+
+	for path, data := range secrets {
+		m.Write(vault, environment, path, data)
+	}
+	return nil
+}
+
+// PullAll pulls all environments
+func (m SecretsManager) PullAll() error {
+	for env := range m.Environments {
+		if err := m.Pull(env); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Pull updates the file on disk with the vaules from Vault (destructive)
+func (m SecretsManager) Pull(environment string) error {
+	vault, err := m.Vault(environment)
+	if err != nil {
+		return err
+	}
+	if _, err := vault.PortForward(); err != nil {
+		return err
+	}
+	defer vault.ClosePortForward()
+
+	p, err := m.GetMapFromVault(vault, environment)
+	if err != nil {
+		return err
+	}
+
+	return m.WriteMapToDisk(vault, environment, p)
+}
+
+// Kubectl creates a Kubectl client
+func (m SecretsManager) Kubectl(environment string) (*KubectlCmd, error) {
+	config := viper.GetStringMap("kubeconfigfiles")
+	directory, ok := config["directory"].(string)
+	if !ok {
+		home, err := homedir.Dir()
+		if err != nil {
+			return nil, err
+		}
+		directory = home + "/.kube/"
+	}
+
+	environments := viper.GetStringMapString("environments")
+
+	return Kubectl(path.Join(directory, environments[environment])), nil
+}
+
+// Vault creates a proxied Vault client
+func (m SecretsManager) Vault(environment string) (*VaultCmd, error) {
```

Comment:
should this be a private function?

Reply:
I thought about it, but I think not, since some of the other functions require a vault instance as an argument.